## Deep Learning - The most powerful and exciting subset of ML
### An Artificial Neural Networks with many layers.

## Artificial Neural Networks - Modeling cognitive behaviour of human mind.

### Problem Statement - We are approached by a MNC Bank who wants to know why people are leaving their bank , the bank has collected a sample of 10k customers and got all their basic details as well as tracked them from 6 months and checked whether they left the bank or all still with the bank.
### Goal - To find out the primary features due to which people are leaving the bank.
### Problem - Classification Problem
### Solution - Through Deep Learning(ANN's) Stochastic Gradient Descent

### Step 1 - Data Preprocessing

In [1]:
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
# Importing the dataset
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:13].values
y = dataset.iloc[:, 13].values
#print(X[0])
#print(type(X[0]))

In [6]:
# Encoding categorical data(As ML model work with numbers and not text)
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X_1 = LabelEncoder()
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])
labelencoder_X_2 = LabelEncoder()
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2])
onehotencoder = OneHotEncoder(categorical_features = [1]) #As there are more than 2 countries(to avoid partiality)
X = onehotencoder.fit_transform(X).toarray()
X = X[:, 1:]

/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_encoders.py:415: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_encoders.py:451: DeprecationWarning: The 'categorical_features' keyword is deprecated in version 0.20 and will be removed in 0.22. You can use the ColumnTransformer instead.
  "use the ColumnTransformer instead.", DeprecationWarning)


In [7]:
# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [8]:
# Feature Scaling , as numbers are of large size increasing performance of the algorithm
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

### Data Preprocessing Code

In [12]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:13].values
y = dataset.iloc[:, 13].values
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X_1 = LabelEncoder()
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1]) #For gender (OHEncoder not required)
labelencoder_X_2 = LabelEncoder()
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2]) #For countries
onehotencoder = OneHotEncoder(categorical_features = [1]) #As there are more than 2 countries(to avoid partiality)
X = onehotencoder.fit_transform(X).toarray()
X = X[:, 1:]
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_encoders.py:415: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_encoders.py:451: DeprecationWarning: The 'categorical_features' keyword is deprecated in version 0.20 and will be removed in 0.22. You can use the ColumnTransformer instead.
  "use the ColumnTransformer instead.", DeprecationWarning)


### Step 2) Building the ANN

In [13]:
# Importing Keras library and required packages
import keras #A library through which we can create neural networks in a few lines of codes
from keras.models import Sequential #Used to initialize our artificial neural network
from keras.layers import Dense #Used to create layers in our artificial neural network
from keras.layers import Dropout #Used to reduce overfitting

Using TensorFlow backend.
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:519: FutureWarn

In [15]:
#Initialzing the ANN
classifier = Sequential()

In [16]:
#Adding the input layer and first hidden layer
#output_dim -  Add a hidden layer with 6 no. nodes(Iip - input(11) + output(1) nodes / 2 = 6 nodes ); init is weights close to 0 ; activ = rectifier; input_dim = no.of independent variables
classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu', input_dim = 11))
classifier.add(Dropout(p = 0.1)) #0.1 means randomly drop 10% of the neurons to avoid overfitting and encourage learning instead of mugging




Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=11, units=6, kernel_initializer="uniform")`
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:4: UserWarning: Update your `Dropout` call to the Keras 2 API: `Dropout(rate=0.1)`
  after removing the cwd from sys.path.


In [17]:
#Adding the second and last hidden layer
classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu',))
classifier.add(Dropout(p = 0.1))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", units=6, kernel_initializer="uniform")`
  
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: UserWarning: Update your `Dropout` call to the Keras 2 API: `Dropout(rate=0.1)`
  This is separate from the ipykernel package so we can avoid doing imports until


In [18]:
#Adding the output layer
classifier.add(Dense(output_dim = 1, init = 'uniform', activation = 'sigmoid',))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="sigmoid", units=1, kernel_initializer="uniform")`
  


In [19]:
#Compiling the ANN
classifier.compile(optimizer = 'adam', loss='binary_crossentropy', metrics = ['accuracy'])



Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where


In [20]:
#Fitting the ANN to the training set
#Batch size - no.of obsv after which you update the weights
#epochs - when whole training set is passed through the ANN and back propagated. (1 iteration of the ANN) and then through cost function and back proprogration weights are updated and 2nd iteration(epoch) begins.
classifier.fit(X_train,y_train,batch_size=10, epochs=100)

Epoch 1/100
8000/8000 [==============================] - 3s 416us/step - loss: 0.4872 - acc: 0.7960
Epoch 2/100
8000/8000 [==============================] - 3s 331us/step - loss: 0.4344 - acc: 0.7960
Epoch 3/100
8000/8000 [==============================] - 3s 327us/step - loss: 0.4313 - acc: 0.7960
Epoch 4/100
8000/8000 [==============================] - 3s 323us/step - loss: 0.4254 - acc: 0.7984
Epoch 5/100
8000/8000 [==============================] - 3s 320us/step - loss: 0.4237 - acc: 0.8187
Epoch 6/100
8000/8000 [==============================] - 2s 224us/step - loss: 0.4259 - acc: 0.8260
Epoch 7/100
8000/8000 [==============================] - 2s 213us/step - loss: 0.4282 - acc: 0.8280
Epoch 8/100
8000/8000 [==============================] - 2s 209us/step - loss: 0.4261 - acc: 0.8275
Epoch 9/100
8000/8000 [==============================] - 2s 212us/step - loss: 0.4225 - acc: 0.8267
Epoch 10/100
8000/8000 [==============================] - 2s 210us/step - loss: 0.4225 - acc: 0.8281

8000/8000 [==============================] - 2s 223us/step - loss: 0.4209 - acc: 0.8322
Epoch 83/100
8000/8000 [==============================] - 2s 225us/step - loss: 0.4200 - acc: 0.8337
Epoch 84/100
8000/8000 [==============================] - 2s 226us/step - loss: 0.4214 - acc: 0.8324
Epoch 85/100
8000/8000 [==============================] - 2s 231us/step - loss: 0.4199 - acc: 0.8340
Epoch 86/100
8000/8000 [==============================] - 2s 222us/step - loss: 0.4227 - acc: 0.8319
Epoch 87/100
8000/8000 [==============================] - 2s 220us/step - loss: 0.4196 - acc: 0.8319
Epoch 88/100
8000/8000 [==============================] - 2s 222us/step - loss: 0.4190 - acc: 0.8334
Epoch 89/100
8000/8000 [==============================] - 3s 356us/step - loss: 0.4167 - acc: 0.8339
Epoch 90/100
8000/8000 [==============================] - 3s 342us/step - loss: 0.4192 - acc: 0.8355
Epoch 91/100
8000/8000 [==============================] - 2s 285us/step - loss: 0.4185 - acc: 0.8332
Epo

### Our ANN model is returning us a probability of 85.99% on training set

##  Step 3 - Making the predictions and evaluating the model

In [21]:
# Predicting the Test set results
#y_pred will contain all probability that the customer will leave the bank
y_pred = classifier.predict(X_test)

In [22]:
1545 + 133
1678 / 2000

0.839

In [15]:
# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

ValueError: Classification metrics can't handle a mix of binary and continuous targets

### 1678 accurate predictions and 322 incorrect predictions; accuracy = no.of correct/ total pred = 83.9 %

### Now the bank can create a ranking based on the probability of customers at the highest risk of leaving the bank, from top to bottom and make strategy to retain them

### Step 4 - Predicting output of a single observation

#### We will manually have to check for one hot encoding for categorical variables from X and dataset matrix
#### [0,0] - Dummy encoding for country 'France'
#### [0] - Female , hence [1] is male written after 600
#### Feature Scaling our data will transform method

In [25]:
new_pred = classifier.predict(sc.transform(np.array([[0, 0 , 600, 1, 40, 3, 60000, 2, 1, 1, 50000]]))) #As ML models take 2d arrays as input
print('The Probability that customer will leave the bank = {} %'.format(round(float(new_pred*100),2)))
new_pred = (new_pred > 0.5)
print('Will the Customer leave the bank ? A] => {}'.format(str(new_pred)))

The Probability that customer will leave the bank = 7.4 %
Will the Customer leave the bank ? A] => [[False]]


### Step 5 - Evaluating, Improving and Tuning our ANN

#### a)Evaluating our ANN (Using k-fold cross validation)
#### (* Execute the data preprocessing step prior to k-fold model analysis)

In [12]:
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import cross_val_score
from keras.models import Sequential
from keras.layers import Dense
def build_classifier():
    classifier = Sequential()
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 11))
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))
    classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
    return classifier
classifier = KerasClassifier(build_fn = build_classifier, batch_size = 10, epochs = 100)
accuracies = cross_val_score(estimator = classifier, X = X_train, y = y_train, cv = 10, n_jobs = -1) #cv = no.of k-folds,n_jobs=-1 means using all CPU's
mean = accuracies.mean()
variance = accuracies.std()
print('Our ANN model is giving us an accuracy of {} %'.format(round(mean*100),3))
print('Variance in our ANN model = {} '.format(round(variance),4))

Our ANN model is giving us an accuracy of 84.0 %
Variance in our ANN model = 0.0 


### Our ANN model is giving us an Relevant accuracy of 84.0 % and a variance of 0.012539935392675127 on the test set
### Challenge - Improving our model's accuracy to 86%

#### b)Improving our ANN 

In [14]:
#Dropout regularization to reduce overfitting

#### c) Tuning our ANN (Finding the Optimum Parameters for our model using GridSearch)